mmseg/models/backbones/vit.py

```diff
@@ -0,0 +1,386 @@
+                    # Upsample pos_embed weights for adapting training inputs.
+                    h, w = self.img_size
+                    pos_embed = VisionTransformer.upsample_pos_embed(
+                        self.pos_embed, h, w, self.pretrain_img_size,
+                        self.pretrain_img_size, self.patch_size)
+                    self.pos_embed = nn.Parameter(pos_embed)
+
+        elif pretrained is None:
+            # We only implement the 'jax_impl' initialization implemented at
+            # https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/vision_transformer.py#L353
+            normal_init(self.pos_embed)
+            for n, m in self.named_modules():
+                if isinstance(m, Linear):
+                    xavier_init(m.weight, distribution='uniform')
+                    if m.bias is not None:
+                        if 'mlp' in n:
+                            normal_init(m.bias, std=1e-6)
+                        else:
+                            constant_init(m.bias, 0)
+                elif isinstance(m, Conv2d):
+                    kaiming_init(m.weight, mode='fan_in')
+                    if m.bias is not None:
+                        constant_init(m.bias, 0)
+                elif isinstance(m, (_BatchNorm, nn.GroupNorm, nn.LayerNorm)):
+                    constant_init(m.bias, 0)
+                    constant_init(m.weight, 1)
+        else:
+            raise TypeError('pretrained must be a str or None')
+
+    def _pos_embeding(self, img, patched_img, pos_embed):
+        """Positiong embeding method.
+
+        Resize the pos_embed, it input image size not equal to training
+            image size.
+        Args:
+            img (torch.Tensor): The inference image tensor, the shape
+                must be [B, C, H, W].
+            patched_img (torch.Tensor): The patched image, it should be
+                shape of [B, L1, C].
+            pos_embed (torch.Tensor): The pos_embed weighs, it should be
+                shape of [B, L2, c].
+        Return:
+            torch.Tensor: The pos encoded image feature.
+        """
+        assert len(patched_img.shape) == 3 and len(pos_embed.shape) == 3, \
+            'the shapes of patched_img and pos_embed must be [B, L, C]'
+        x_len, pos_len = patched_img.shape[1], pos_embed.shape[1]
+        if x_len != pos_len:
+            # If run tools/test.py to test a model, the shape of pos_embed
+            # should be [B, L_pretrain, C]
+            if pos_len == (self.pretrain_img_size // self.patch_size)**2:
+                pos_h, pos_w = self.pretrain_img_size, self.pretrain_img_size
+            # During evaluation at training, the shape of pos_embed should be
+            # [B, L_training, C]
+            elif pos_len == (self.img_size[0] // self.patch_size) * (
+                    self.img_size[1] // self.patch_size):
+                pos_h, pos_w = self.img_size
```

Comment:
These cases could be simplified. 
We need two types of pos_embeding. 
Type 1: Fine-tuning from pretrained backbone. We may infer the new shape of pos embedding from `self.img_size`. 
Type 2: Vanilla forward. No matter training or evaluation, we will resize the pos embedding according to the input image size.